### ISSUE TITLE
Bug: `F821` Error for complex import of `Literal`

### ISSUE THREAD

**Issue by code-yeongyu** [astral-sh/ruff#6651]
Hey there! 👋

First off, thank you for the ruff project. It's my second time to report an issue. I love your product and enjoy using it for everyday.

If i import of `Literal` in a complex way, it raises F821.

Here is the minimal code snippet to reproduce the bug.

The following code raises F821 of "Undefined name `greeting`" on `main.py`

`custom_types.py`:
```python
from typing import Literal as JustString
```

`main.py`:
```python
from custom_types import JustString

choice: JustString["greeting"]
```

Thank you.

**Comment by charliermarsh** [comment]
Thanks for the kind words! Unfortunately this is a known issue. It'll be fixed at some point but it's not quite trivial...

If you consistently used the name `Literal` instead of aliasing to `JustString`, you could use [`typing-modules`](https://beta.ruff.rs/docs/settings/#typing-modules):

```toml
[tool.ruff]
typing-modules = ["custom_types"]
```

Then, when Ruff sees `from custom_types import Literal`, it would interpret that as equivalent to `typing.Literal`. But if the import is renamed in `custom_types`, we aren't able to trace that `JustString` is equivalent to `Literal` (yet).

**Comment by charliermarsh** [comment]
(Note: this requires multi-file support, since we need to resolve `JustString` as equivalent to `Literal` across files.)

**Comment by code-yeongyu** [comment]
Ah I see. Seems it's very complex issue. Just like I have reported issue before, thank you for your speedy response and kindness. Have a good day!

**Comment by charliermarsh** [comment]
I think this can be merged with #6014.
